src/Symfony/Component/Form/Tests/Extension/Validator/Constraints/FormValidatorFunctionalTest.php

```diff
@@ -0,0 +1,162 @@
+        $this->assertSame('This value should not be blank.', $violations[0]->getMessage());
+        $this->assertSame('data[field1]', $violations[0]->getPropertyPath());
+        $this->assertSame('This value should not be blank.', $violations[1]->getMessage());
+        $this->assertSame('data[field2]', $violations[1]->getPropertyPath());
+    }
+
+    public function testCompositeConstraintValidatedInSequence()
+    {
+        $form = $this->formFactory->create(FormType::class, null, [
+            'constraints' => [
+                new Collection([
+                    'field1' => new NotBlank([
+                        'groups' => ['field1'],
+                    ]),
+                    'field2' => new NotBlank([
+                        'groups' => ['field2'],
+                    ]),
+                ]),
+            ],
+            'validation_groups' => new GroupSequence(['field1', 'field2']),
+        ]);
+        $form->add('field1');
+        $form->add('field2');
+
+        $form->submit([
+            'field1' => '',
+            'field2' => '',
+        ]);
+
+        $violations = $this->validator->validate($form);
+
+        $this->assertCount(1, $violations);
+        $this->assertSame('This value should not be blank.', $violations[0]->getMessage());
+        $this->assertSame('data[field1]', $violations[0]->getPropertyPath());
+    }
+
+    public function testCascadeValidationToChildFormsUsingPropertyPaths()
+    {
+        $form = $this->formFactory->create(FormType::class, null, [
+            'validation_groups' => ['group1', 'group2'],
+        ])
+            ->add('field1', null, [
+                'constraints' => [new NotBlank(['groups' => 'group1'])],
+                'property_path' => '[foo]',
+            ])
+            ->add('field2', null, [
+                'constraints' => [new NotBlank(['groups' => 'group2'])],
+                'property_path' => '[bar]',
+            ])
+        ;
+
+        $form->submit([
+            'field1' => '',
+            'field2' => '',
+        ]);
+
+        $violations = $this->validator->validate($form);
+
+        $this->assertCount(2, $violations);
+        $this->assertSame('This value should not be blank.', $violations[0]->getMessage());
+        $this->assertSame('children[field1].data', $violations[0]->getPropertyPath());
+        $this->assertSame('This value should not be blank.', $violations[1]->getMessage());
+        $this->assertSame('children[field2].data', $violations[1]->getPropertyPath());
+    }
+
+    public function testCascadeValidationToChildFormsUsingPropertyPathsValidatedInSequence()
+    {
+        $form = $this->formFactory->create(FormType::class, null, [
+            'validation_groups' => new GroupSequence(['group1', 'group2']),
+        ])
+            ->add('field1', null, [
+                'constraints' => [new NotBlank(['groups' => 'group1'])],
+                'property_path' => '[foo]',
+            ])
+            ->add('field2', null, [
```

Comment:
While working on #37085 I realised that we couldn't add this second field and assert its constraints are not validated. Investigating the cause for it I found out that the default `ConstraintValidatorTestCase` is not well suited for these kinds of tests where we cascade validation to nested fields.